xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -170,6 +187,59 @@ void watchConfigData(String hostName, int port, ConfigWatcher watcher) {
     adsStream.sendXdsRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName));
   }
 
+  @Override
+  void watchEndpointData(String clusterName, EndpointWatcher watcher) {
+    boolean needRequest = false;
+    if (!endpointWatchers.containsKey(clusterName)) {
+      logger.log(Level.FINE, "Start watching endpoints in cluster {0}", clusterName);
+      needRequest = true;
+      endpointWatchers.put(clusterName, new HashSet<EndpointWatcher>());
+    }
+    Set<EndpointWatcher> watchers = endpointWatchers.get(clusterName);
+    if (watchers.contains(watcher)) {
+      logger.log(Level.WARNING, "Watcher {0} already registered", watcher);
+    }
+    watchers.add(watcher);
+    // If local cache contains endpoint information for the cluster to be watched, notify
+    // the watcher immediately.
+    if (clusterNamesToEndpointUpdates.containsKey(clusterName)) {
+      watcher.onEndpointChanged(clusterNamesToEndpointUpdates.get(clusterName));
+    }
+    if (rpcRetryTimer != null) {
+      // Currently in retry backoff.
+      return;
+    }
+    if (needRequest) {
+      if (adsStream == null) {
+        startRpcStream();
+      }
+      adsStream.sendXdsRequest(ADS_TYPE_URL_EDS, endpointWatchers.keySet());
+    }
+  }
+
+  @Override
+  void cancelEndpointDataWatch(String clusterName, EndpointWatcher watcher) {
+    Set<EndpointWatcher> watchers = endpointWatchers.get(clusterName);
+    if (watchers == null) {
+      logger.log(Level.WARNING, "Watcher {0} was not registered", watcher);
+      return;
+    }
+    watchers.remove(watcher);
```

Comment:
if this remove returned `false` you should generate the same error "Watcher {0} was not registered" and return.

Reply:
Done. Added in the previous `if` to include this case.